sbat.c

```diff
@@ -335,4 +335,62 @@ parse_sbat_var(list_t *entries)
 	FreePool(data);
 	return EFI_INVALID_PARAMETER;
 }
+
+EFI_STATUS
+init_sbat()
+{
+	EFI_STATUS efi_status = EFI_SUCCESS;
+	UINT32 attributes = 0;
+
+	CHAR8 *sbat;
+	UINTN sbatsize = 4096;
+
+	sbat = AllocateZeroPool(sbatsize);
+
+	/* read and get check attributes */
+	efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID, &attributes,
+			&sbatsize, sbat);
+
+	/*
+	 * If attributes and the initial part of the version match, then
+	 * there is no need to re-write it. If it is not set or the read
+	 * failed, then attributes won't match.
+	 */
+	if (((attributes != (EFI_VARIABLE_NON_VOLATILE |
+		EFI_VARIABLE_BOOTSERVICE_ACCESS)) ||
+		(attributes != (EFI_VARIABLE_NON_VOLATILE |
+				EFI_VARIABLE_BOOTSERVICE_ACCESS |
+				EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS)))
+			|| (strncmp(sbat, SBAT_INIT, sizeof(SBAT_SIG) - 1) != 0)) {
+
+		efi_status = gRT->SetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				attributes, 0, "");
+
+		/* verify that it's gone */
+		ZeroMem(sbat, sbatsize);
+		efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				0, &sbatsize, sbat);
+
+		/* set variable */
+		efi_status = gRT->SetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				EFI_VARIABLE_NON_VOLATILE |
+				EFI_VARIABLE_BOOTSERVICE_ACCESS,
+				sizeof(SBAT_INIT), SBAT_INIT);
+
+		/* verify that the expected data is there*/
+		ZeroMem(sbat, sbatsize);
+		efi_status = gRT->GetVariable(L"SBAT", &SHIM_LOCK_GUID,
+				0, &sbatsize, sbat);
+		if (strncmp(sbat, SBAT_INIT, sizeof(SBAT_INIT)) != 0) {
```

Comment:
Pretty sure you want strlen() instead of sizeof() here; see https://gist.github.com/vathpela/cb118cd7d623ce8022107fea784e6556

Reply:
Good catch!
Adding a manual test case where the next char doesn't match as well.